File: ch3/dp/beepers_UVa10496.cpp
// Collecting Beepers

#include <bits/stdc++.h>
using namespace std;

#define LSOne(S) ((S) & -(S))

const int MAX_n = 11;

int dist[MAX_n][MAX_n], memo[MAX_n][1<<(MAX_n-1)]; // Karel + max 10 beepers

int dp(int u, int mask) {                        // mask = free coordinates
  if (mask == 0) return dist[u][0];              // close the loop
  int &ans = memo[u][mask];
  if (ans != -1) return ans;                     // computed before
  ans = 2000000000;
  int m = mask;
  while (m) {                                    // up to O(n)
    int two_pow_v = LSOne(m);                    // but this is fast
    int v = __builtin_ctz(two_pow_v)+1;          // offset v by +1
    ans = min(ans, dist[u][v] + dp(v, mask^two_pow_v)); // keep the min
    m -= two_pow_v;
  }
  return ans;
}

int main() {
  int TC; scanf("%d", &TC);
  while (TC--) {
    int xsize, ysize; scanf("%d %d", &xsize, &ysize); // these two values are not used
    int x[MAX_n], y[MAX_n];
    scanf("%d %d", &x[0], &y[0]);
    int n; scanf("%d", &n); ++n;                 // include Karel
    for (int i = 1; i < n; ++i)                  // Karel is at index 0
      scanf("%d %d", &x[i], &y[i]);
    for (int i = 0; i < n; ++i)                  // build distance table
      for (int j = i; j < n; ++j)
        dist[i][j] = dist[j][i] = abs(x[i]-x[j]) + abs(y[i]-y[j]); // Manhattan distance
    memset(memo, -1, sizeof memo);
    printf("%d\n", dp(0, (1<<(n-1))-1));         // DP-TSP
    // printf("The shortest path has length %d\n", dp(0, (1<<(n-1))-1)); // DP-TSP
  }
  return 0;
}

File: ch3/dp/beepers_UVa10496.py
import sys,math
from functools import lru_cache
sys.setrecursionlimit(5000)

def LSOne(S):
    return ((S) & -(S))

def main():
    TC = int(input())
    for _ in range(TC):
        input()                                  # these two values: xsize ysize are not used
        x = [0]*11
        y = [0]*11
        x[0],y[0] = map(int, input().split())
        n = int(input())
        n += 1                                   # include Karel
        for i in range(1, n):                    # Karel is at index 0
            x[i],y[i] = map(int, input().split())
        dist = [[0] * n for _ in range(n)]       # Karel + max 10 beepers
        for i in range(n-1):                     # build distance table
            for j in range(i+1, n):
                dist[i][j] = dist[j][i] = abs(x[i]-x[j]) + abs(y[i]-y[j]) # Manhattan distance

        @lru_cache(maxsize=None)
        def dp(u, mask):                         # mask = free coordinates
            if mask == 0: return dist[u][0]      # close the loop
            ans = 2000000000
            m = mask
            while (m):                           # up to O(n)
                two_pow_v = LSOne(m)             # but this is fast
                v = int(math.log2(two_pow_v))+1  # offset v by +1
                ans = min(ans, dist[u][v] + dp(v, mask^two_pow_v)) # keep the min
                m -= two_pow_v
            return ans

        print(dp(0, (1<<(n-1))-1))         # DP-TSP
        # print("The shortest path has length {}".format(dp(0, (1<<(n-1))-1))) # DP-TSP

main()
